docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,753 @@
+     */
+    AsyncResponseTransformer<GetObjectResponse, ?> transformerForPart(MultipartDownloadContext context);
+
+    /**
+     * Return a transformer for downloading an entire object as a single part.
+     */
+    AsyncResponseTransformer<GetObjectResponse, ?> transformerForObject(SinglePartDownloadContext context)
+
+    /**
+     * Return a factory capable of creating transformers that will recombine the
+     * object parts to a single file on disk.
+     */
+    static ObjectPartTransformerCreator forFile(Path file) {
+        return ...;
+    }
+}
+
+/**
+ * The context object for the upload of an object part to S3.
+ */
+public interface MultipartUploadContext {
+    /**
+     * The original upload request given to the transfer manager.
+     */
+    UploadObjectRequest uploadRequest();
+
+    /**
+     * The request sent to S3 to initiate the multipart upload.
+     */
+    CreateMultipartUploadRequest createMultipartRequest();
+
+    /**
+     * The upload request to be sent to S3 for this part.
+     */
+    UploadPartRequest uploadPartRequest();
+
+    /**
+     * The offset from the beginning of the object where this part begins.
+     */
+    long partOffset();
+}
+
+public interface SinglePartUploadContext {
+    /**
+     * The original upload request given to the transfer manager.
+     */
+    UploadObjectRequest uploadRequest();
+
+    /**
+     * The request to be sent to S3 to upload the object as a single part.
+     */
+    PutObjectRequest objectRequest();
+}
+
+/**
+ * Context object for an individual object part for a multipart download.
+ */
+public interface MultipartDownloadContext {
+    /**
+     * The original download request given to the Transfer Manager.
+     */
+    DownloadObjectRequest downloadRequest();
+
+    /**
+     * The request sent to S3 for this object part. This is empty if downloading a presigned URL.
+     */
+    Optional<GetObjectRequest> partRequest();
+
+    /**
+     * The part number.  */
+    int partNumber();
+
+    /**
+     * The offset from the beginning of the object where this part begins.
+     */
+    long partOffset();
+
+    /**
+     * The size of the part requested in bytes.
+     */
+    long size();
+
+    /**
+     * Whether this is the last part of the object.
+     */
+    boolean isLastPart();
+}
+
+/**
+ * Context object for a single part download of an object.
+ */
+public interface SinglePartDownloadContext {
+    /**
+     * The original download request given to the Transfer Manager.
+     */
+    DownloadObjectRequest downloadRequest();
+
+    /**
+     * The request sent to S3 for this object. This is empty if downloading a presigned URL.
+     */
+    Optional<GetObjectRequest> objectRequest();
+}
+
+/**
+ * Progress listener for a Transfer.
+ * <p>
+ * The SDK guarantees that calls to {@link #transferProgressEvent(EventContext)}
+ * are externally synchronized.
+ */
+public interface TransferProgressListener {
```

Comment:
When are we going to figure out how this relates to the progress listeners in the SDK core?

Reply:
I wanted to make sure we could support progress events in the way we want to for TransferManager first, then ideally we could generalize that for the core listeners.

Reply:
Will we do that before we GA transfer-manager?

Reply:
Yes definitely